test/e2e/storage/host_path_type.go

```diff
@@ -0,0 +1,436 @@
+		hostPathSocket            = v1.HostPathSocket
+		hostPathCharDev           = v1.HostPathCharDev
+		hostPathBlockDev          = v1.HostPathBlockDev
+	)
+
+	ginkgo.BeforeEach(func() {
+		ns = f.Namespace.Name
+		node, err := e2enode.GetRandomReadySchedulableNode(f.ClientSet)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Create a pod for further testing")
+		ginkgo.By(fmt.Sprintf("running on node %s", node.Name))
+		hostBaseDir = path.Join("/tmp", ns)
+		mountBaseDir = "/mnt/test"
+		basePod = f.PodClient().CreateSync(newHostPathTypeTestPod(node.Name, hostBaseDir, mountBaseDir, &hostPathDirectoryOrCreate))
+
+		targetBlockDev = path.Join(hostBaseDir, "ablkdev")
+		ginkgo.By("Create a block device for further testing")
+		_, err = utils.PodExec(f, basePod, fmt.Sprintf("mknod %s b 89 1", path.Join(mountBaseDir, "ablkdev")))
+		framework.ExpectNoError(err)
+	})
+
+	ginkgo.It("Should fail on mounting non-existent block device 'does-not-exist-blk-dev' when HostPathType is HostPathBlockDev [Slow]", func() {
+		blkDevPath := path.Join(hostBaseDir, "does-not-exist-blk-dev")
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, blkDevPath, fmt.Sprintf("%s is not a block device", blkDevPath), &hostPathBlockDev)
+	})
+
+	ginkgo.It("Should be able to mount block device 'ablkdev' successfully when HostPathType is HostPathBlockDev", func() {
+		verifyPodHostPathType(f, basePod.Spec.NodeName, targetBlockDev, &hostPathBlockDev)
+	})
+
+	ginkgo.It("Should be able to mount block device 'ablkdev' successfully when HostPathType is HostPathUnset", func() {
+		verifyPodHostPathType(f, basePod.Spec.NodeName, targetBlockDev, &hostPathUnset)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathDirectory [Slow]", func() {
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a directory", targetBlockDev), &hostPathDirectory)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathFile [Slow]", func() {
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a file", targetBlockDev), &hostPathFile)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathSocket [Slow]", func() {
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a socket", targetBlockDev), &hostPathSocket)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathCharDev [Slow]", func() {
+		verifyPodHostPathTypeFailure(f, basePod.Spec.NodeName, targetBlockDev, fmt.Sprintf("%s is not a character device", targetBlockDev), &hostPathCharDev)
+	})
+})
+
+func newHostPathTypeTestPod(nodeName, hostDir, mountDir string, hostPathType *v1.HostPathType) *v1.Pod {
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-hostpath-type-",
+		},
+		Spec: v1.PodSpec{
+			NodeName:      nodeName,
```

Comment:
We had an issue recently where nodes got overloaded and tests failed because kubelet immediately rejected the Pod with "out of pods". https://github.com/kubernetes/kubernetes/issues/87855

Can we use NodeSelector on hostname label here? That way the Pod will still go through the scheduler and stay pending until the node gets freed up.